### PR TITLE
Faster deserialisation of still experiments via information reuse

### PR DESCRIPTION
**Pull request by ndevenish**
This dramatically increases performance when deserialising experiment lists constructed as one-imageset-per-image, as they currently commonly appear to be[1] for still image sets. This reduces the load time to around ~5s for an example 11k image dataset. This is down from *too long to measure*, which is unsurprising given the "read every image for every image" loops it removes.

Some of the changes here aren't elegant solutions - there are some severe restrictions imposed by the data model and the cyclic nature of needing to resolve the format to construct an imageset in order to deserialize an imageset. These should be confronted, but at a later date.

I believe this overlaps with some of the 'Lazy' and 'load_models' solutions put in in the recent past:
- https://github.com/cctbx/cctbx_project/pull/313 (Add option to delay load models when reading raw data)
- https://github.com/cctbx/cctbx_project/pull/312 (Lazy experiment list)
- https://github.com/dials/dials/pull/802 (Updates to dxtbx and dials to allow image viewer regarding load_models)
though I'm not certain which stage of the problem all of those are dealing with so it might not overlap all of them.

Changes made:
- `FormatMultiImage.get_imageset` no longer attempts to read every model for every image if a beam and detector has been provided and a single file index has been requested. This prevents the worst of the N² file reading loops that was preventing still experiment lists from loading
- `FormatMultiImage.get_imageset` caches the format instance from the last time it was called. If it is immediately asked for the same filename and format arguments again, it reuses these instances rather than trying to recreate the format class, which did a lot of file IO
- `FormatMultiImage` also reuses `ImageSetData` objects from the last time it was called - otherwise you end up with N unique `ImageSetData` objects allocating 8²×N² bytes of pointers (7.3 GB for 11k images!)
- `ExperimentListDict._make_stills` also caches previous filename-format lookups so that this doesn't need to be redone for every image in a multi-image file.

[1]: This isn't completely true, as there are tests like [this](https://github.com/cctbx/dxtbx/commit/381b90fbcc07e8889623a5ca39beaae00fae0aea#diff-46bae859cf3345bbbc948e5b7e463da2R776) which explicitly load multiple-image still ImageSet objects. I don't think it's used in live analysis at the moment, though we have been discussing changing this.

### CODE REVIEW

**Comment by ndevenish** [comment]
To add a clarification: I've been testing with single-file datasets e.g. nexus. Not sure how single-file datasets will fair but many of the problems were specific to `FormatMultiImage` here.

File: format/FormatMultiImage.py [hunk 1]
```diff
@@ -263,40 +287,56 @@ def get_imageset(
...
-                iset.set_scan(scan[single_file_indices[i]], i)
+            else:
+                # Asking for multiple indices AND providing a model makes no sense
+                assert len(single_file_indices) == 1 or (not beam and not detector)
+
+            if len(single_file_indices) == 1:
+                single_file_index = single_file_indices[0]
+                # If we don't have a beam and detector provided we need
+                # to read it - but only for this image
+                if all(x is None for x in (beam, detector)):
```
Comment (Anthchirp) [inline format/FormatMultiImage.py]:
```suggestion
                if not beam and not detector:
```

File: model/experiment_list.py [hunk 2]
```diff
@@ -371,22 +381,33 @@ def _make_mem_imageset(self, imageset):
         """ Can't make a mem imageset from dict. """
         return None
 
-    def _make_stills(self, imageset, format_kwargs=None):
-        """ Make a still imageset. """
+    def _make_stills(self, imageset, beam=None, detector=None, format_kwargs=None):
+        """ Make a still imageset."""
```
Comment (Anthchirp) [inline model/experiment_list.py]:
```suggestion
        """Make a still imageset."""
```

File: model/experiment_list.py [hunk 3]
```diff
@@ -371,22 +381,33 @@ def _make_mem_imageset(self, imageset):
...
         indices = None
         if "single_file_indices" in imageset:
             indices = imageset["single_file_indices"]
             assert len(indices) == len(filenames)
-        return ImageSetFactory.make_imageset(
+
+        # Do we know the format class? Pass it through if we do to save time
+        format_class = None
+        if filenames[0] in self._imageset_format_cache:
+            format_class = self._imageset_format_cache[filenames[0]]
```
Comment (Anthchirp) [inline model/experiment_list.py]: replace 3 lines with
```python
format_class = self._imageset_format_cache.get(filenames[0])
```

File: format/FormatMultiImage.py [hunk 4]
```diff
@@ -193,9 +208,18 @@ def get_imageset(
         # Get the format instance
         assert len(filenames) == 1
         if check_format is True:
```
Comment (Anthchirp) [inline format/FormatMultiImage.py]:
```suggestion
         if check_format:
```

**Pull request merged**